Compilation issue: UINT64_MAX/UINT64_C not declared in scope

R 3.6.0
dqrng 0.2.0 
Centos 6 with gcc 5.3.0

I encountered the following compilation issue:

```In file included from ../inst/include/dqrng_generator.h:25:0,
                 from dqrng.cpp:19:
../inst/include/xoshiro.h: In static member function ‘static constexpr dqrng::xoshiro<N, A, B, C>::result_type dqrng::xoshiro<N, A, B, C>::max()’:
../inst/include/xoshiro.h:80:53: error: ‘UINT64_MAX’ was not declared in this scope
   inline static constexpr result_type max() {return UINT64_MAX;};
                                                     ^
../inst/include/xoshiro.h: In member function ‘void dqrng::xoshiro<N, A, B, C>::jump() [with long unsigned int N = 2ul; signed char A = 24; signed char B = 16; signed char C = 37]’:
../inst/include/xoshiro.h:144:31: error: ‘UINT64_C’ was not declared in this scope
       if (JUMP[i] & UINT64_C(1) << b) {
```

Adding the following lines to inst/include/xoshiro.h (after the #include cstdint statement) fixes the issue:

```
#ifndef UINT64_MAX
#define UINT64_MAX ((uint64_t) -1)
#endif

#ifndef UINT64_C
#define UINT64_C(c) c
#endif
```

--- a/inst/include/dqrng_distribution.h
+++ b/inst/include/dqrng_distribution.h
@@ -18,7 +18,7 @@
 #ifndef DQRNG_DISTRIBUTION_H
 #define DQRNG_DISTRIBUTION_H 1
 
-#include <cstdint>
+#include <mystdint.h>
 #include <memory>
 #include <boost/random/uniform_real_distribution.hpp>
 #include <boost/random/normal_distribution.hpp>
@@ -34,7 +34,7 @@ using exponential_distribution = boost::random::exponential_distribution<double>
 
 inline double uniform01(uint64_t x) {
   // prefer high bits due to weakness of lowest bits for xoshiro/xoroshiro with used "+" scrambler
-  return (x >> 11) * (1. / (UINT64_C(1) << 53));
+  return (x >> 11) * (1. / (uint64_t(1) << 53));
 }
 } // namespace dqrng
 

--- a/inst/include/dqrng_generator.h
+++ b/inst/include/dqrng_generator.h
@@ -18,7 +18,7 @@
 #ifndef DQRNG_GENERATOR_H
 #define DQRNG_GENERATOR_H 1
 
-#include <cstdint>
+#include <mystdint.h>
 #include <memory>
 #include <type_traits>
 #include <stdexcept>

--- a/inst/include/mystdint.h
+++ b/inst/include/mystdint.h
@@ -1,0 +1,34 @@
+#ifndef MYSTDINT_H
+#define MYSTDINT_H 1
+
+/* C99 defined macros like UINT64_MAX and UINT64_C but made them unavailable
+ * to C+x unless __STDC_LIMIT_MACROS and __STDC_CONSTANT_MACROS are defined.
+ * This was removed with C++11 and C11, but some old libc implementations still
+ * have that (e.g. the one used in RHEL/CentOS/... 6). C++ compilers from that
+ * time worked around this issue in the cstdint header, but newer compilers
+ * don't do that any more. This file reintroduces these tricks to support old
+ * libc versions together with newer compilers.
+ *
+ * Limitation: This header has no effect if cstdint is included before it.
+ */
+
+#ifndef __STDC_LIMIT_MACROS
+# define _UNDEF__STDC_LIMIT_MACROS
+# define __STDC_LIMIT_MACROS
+#endif
+#ifndef __STDC_CONSTANT_MACROS
+# define _UNDEF__STDC_CONSTANT_MACROS
+# define __STDC_CONSTANT_MACROS
+#endif
+#include <cstdint>
+#ifdef _UNDEF__STDC_LIMIT_MACROS
+# undef __STDC_LIMIT_MACROS
+# undef _UNDEF__STDC_LIMIT_MACROS
+#endif
+#ifdef _UNDEF__STDC_CONSTANT_MACROS
+# undef __STDC_CONSTANT_MACROS
+# undef _UNDEF__STDC_CONSTANT_MACROS
+#endif
+
+
+#endif // MYSTDINT_H

--- a/inst/include/mystdint.h
+++ b/inst/include/mystdint.h
@@ -30,5 +30,10 @@
 # undef _UNDEF__STDC_CONSTANT_MACROS
 #endif
 
+// define our own UINT64_MAX in case it is still missing
+#ifndef UINT64_MAX
+#include <limits>
+#define UINT64_MAX std::numeric_limits<std::uint64_t>::max()
+#endif
 
 #endif // MYSTDINT_H

--- a/inst/include/xoshiro.h
+++ b/inst/include/xoshiro.h
@@ -12,7 +12,7 @@
 #define XOSHIRO_H 1
 
 #include <array>
-#include <cstdint>
+#include <mystdint.h>
 #include <functional>
 #include <algorithm>
 #include <type_traits>

--- a/inst/include/xoshiro.h
+++ b/inst/include/xoshiro.h
@@ -144,7 +144,7 @@ inline void xoroshiro128plus::jump() {
   uint64_t s1 = 0;
   for(unsigned int i = 0; i < sizeof JUMP / sizeof *JUMP; i++)
     for(unsigned int b = 0; b < 64; b++) {
-      if (JUMP[i] & UINT64_C(1) << b) {
+      if (JUMP[i] & uint64_t(1) << b) {
         s0 ^= state[0];
         s1 ^= state[1];
       }
@@ -167,7 +167,7 @@ inline void xoroshiro128plus::long_jump(void) {
 	uint64_t s1 = 0;
 	for(unsigned int i = 0; i < sizeof LONG_JUMP / sizeof *LONG_JUMP; i++)
 		for(unsigned int b = 0; b < 64; b++) {
-			if (LONG_JUMP[i] & UINT64_C(1) << b) {
+			if (LONG_JUMP[i] & uint64_t(1) << b) {
 				s0 ^= state[0];
 				s1 ^= state[1];
 			}
@@ -208,7 +208,7 @@ inline void xoshiro256plus::jump() {
   uint64_t s3 = 0;
   for(unsigned int i = 0; i < sizeof JUMP / sizeof *JUMP; i++)
     for(unsigned int b = 0; b < 64; b++) {
-      if (JUMP[i] & UINT64_C(1) << b) {
+      if (JUMP[i] & uint64_t(1) << b) {
         s0 ^= state[0];
         s1 ^= state[1];
         s2 ^= state[2];
@@ -237,7 +237,7 @@ inline void xoshiro256plus::long_jump(void) {
 	uint64_t s3 = 0;
 	for(unsigned int i = 0; i < sizeof LONG_JUMP / sizeof *LONG_JUMP; i++)
 		for(unsigned int b = 0; b < 64; b++) {
-			if (LONG_JUMP[i] & UINT64_C(1) << b) {
+			if (LONG_JUMP[i] & uint64_t(1) << b) {
 				s0 ^= state[0];
 				s1 ^= state[1];
 				s2 ^= state[2];

--- a/src/dqrng.cpp
+++ b/src/dqrng.cpp
@@ -15,6 +15,7 @@
 // You should have received a copy of the GNU Affero General Public License
 // along with dqrng.  If not, see <http://www.gnu.org/licenses/>.
 
+#include <mystdint.h>
 #include <Rcpp.h>
 #include <dqrng_generator.h>
 #include <dqrng_distribution.h>

